File: app/cmd/nodes.go
// 文件路径: WMB/app/cmd/nodes.go

package main

import (
	"fmt"

	"github.com/spf13/cobra"
	"github.com/wumindang/WMB/app/nodes" // 替换为实际模块路径
)

var nodeManager = nodes.NewNodeManager()

var nodesCmd = &cobra.Command{
	Use:   "nodes",
	Short: "管理区块链节点",
	Long:  `添加、移除或列出区块链网络中的节点。`,
}

var addNodeCmd = &cobra.Command{
	Use:   "add [peer_id] [address]",
	Short: "添加一个节点",
	Long:  `向节点管理器添加一个新的节点，指定 PeerID 和 Multiaddr。`,
	Args:  cobra.ExactArgs(2),
	Run: func(cmd *cobra.Command, args []string) {
		peerID := args[0]
		address := args[1]

		err := nodeManager.AddNode(peerID, address)
		if err != nil {
			fmt.Println("错误:", err)
			return
		}
		fmt.Println("节点添加成功:", peerID)
	},
}

var removeNodeCmd = &cobra.Command{
	Use:   "remove [peer_id]",
	Short: "移除一个节点",
	Long:  `从节点管理器中移除指定 PeerID 的节点。`,
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		peerID := args[0]

		err := nodeManager.RemoveNode(peerID)
		if err != nil {
			fmt.Println("错误:", err)
			return
		}
		fmt.Println("节点移除成功:", peerID)
	},
}

var listNodesCmd = &cobra.Command{
	Use:   "list",
	Short: "列出所有节点",
	Long:  `获取当前节点管理器中的所有节点信息。`,
	Run: func(cmd *cobra.Command, args []string) {
		jsonStr, err := nodeManager.ToJSON()
		if err != nil {
			fmt.Println("错误:", err)
			return
		}
		fmt.Println("节点列表:", jsonStr)
	},
}

func init() {
	nodesCmd.AddCommand(addNodeCmd)
	nodesCmd.AddCommand(removeNodeCmd)
	nodesCmd.AddCommand(listNodesCmd)
	rootCmd.AddCommand(nodesCmd)
}

File: src/nodes.rs
// 文件路径: src/nodes.rs
// 作用: 管理不同类型的节点

use serde::{Serialize, Deserialize}; // 序列化和反序列化支持
use serde_json;                     // JSON 序列化支持
use std::collections::HashMap;       // 哈希表
use std::sync::{Arc, Mutex};         // 线程安全锁

// 节点类型枚举
#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum NodeType {
    CitizenLight,       // 公民轻节点
    GuestLight,         // 访客轻节点
    Super,              // 超级节点
    NationalAuthority,  // 国储会权威节点
    ProvincialAuthority,// 省储会权威节点
    ProvincialEquity,   // 省储行权益节点
    Full,               // 全节点
}

// 节点结构
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Node {
    pub peer_id: String,    // 节点 Peer ID
    pub address: String,    // 节点地址（Multiaddr）
    pub status: String,     // 节点状态（例如 "active"）
    pub node_type: NodeType,// 节点类型
}

// 节点管理器
#[derive(Clone)]
pub struct NodeManager {
    nodes: Arc<Mutex<HashMap<String, Node>>>, // 线程安全的节点哈希表
}

impl NodeManager {
    // 创建新节点管理器
    pub fn new() -> Self {
        NodeManager {
            nodes: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    // 添加节点
    pub fn add_node(&self, peer_id: &str, address: &str, node_type: NodeType) -> Result<(), String> {
        let mut nodes = self.nodes.lock().map_err(|e| format!("锁失败: {}", e))?; // 获取锁
        if nodes.contains_key(peer_id) {
            return Err(format!("节点 {} 已存在", peer_id)); // 检查是否重复
        }
        nodes.insert(
            peer_id.to_string(),
            Node {
                peer_id: peer_id.to_string(),
                address: address.to_string(),
                status: "active".to_string(),
                node_type,
            },
        ); // 添加新节点
        Ok(())
    }

    // 移除节点
    pub fn remove_node(&self, peer_id: &str) -> Result<(), String> {
        let mut nodes = self.nodes.lock().map_err(|e| format!("锁失败: {}", e))?; // 获取锁
        if nodes.remove(peer_id).is_none() {
            return Err(format!("节点 {} 不存在", peer_id)); // 检查是否存在
        }
        Ok(())
    }

    // 列出所有节点
    pub fn list_nodes(&self) -> Result<Vec<Node>, String> {
        let nodes = self.nodes.lock().map_err(|e| format!("锁失败: {}", e))?; // 获取锁
        Ok(nodes.values().cloned().collect()) // 返回节点列表
    }

    // 将节点列表转换为 JSON
    pub fn to_json(&self) -> Result<String, String> {
        let nodes = self.list_nodes()?; // 获取节点列表
        serde_json::to_string(&nodes).map_err(|e| format!("JSON 序列化失败: {}", e)) // 转换为 JSON
    }
}